### ISSUE TITLE
Adding route specific middleware?

### ISSUE THREAD

**Issue by rw12345** [apigee-127/swagger-tools#291]
Hi There,
Am I missing something or it's not currently possible to add a route specific middleware?

Let's say I have an API /users that's used to create a user. 
Swagger tools makes it insanely easy to declare the spec, and establish the relationship [1] with a controller that receives the validated params [2] to create the user. So far, so good.

As an additional security requirement, I would like to have a middleware that checks if the request for user creation comes from an allowed range of IP addresses. 

In standard express speak, I'd have had a route middleware on the user post request handler shown in [3].

How do I accomplish the same using swagger tools? Looking at https://github.com/apigee-127/swagger-tools/blob/master/docs/Middleware.md, it seems that middleware handler is all or nothing and abstracts away the routes completely.

Thanks!

[1] Swagger definition
```
paths:
  /users:
    x-swagger-router-controller: userController
    post:
      summary: Creates a new user
      description:
        Creates a new user with the given name.
      operationId: post
      parameters:
        - in: body
          name: body
          description: The user object
          required: true
          schema:
            $ref: "#/definitions/User"
...
```

[2] userController.js
```
function post(req, res, next) {
  var body = req.swagger.params.body.value;
  // Create user here...
}
module.exports = {
  post: post
...
```

[3] Simple express app with user create route and IP address checking middleware
```
var express = require('express');
var app = express();

// Middleware function for IP address whitelist checking
var function checkIPAddressAllowed = function(req, res, next) {
  // Return next with error to disallow
  ...
};

var createUser = function(req, res, next) {
  // Create user
  ...
};

app.post('/users', checkIPAddressAllowed, createUser);
```

**Comment by whitlockjc** [comment]
I don't think swagger-tools is stopping you from doing this but since it depends on your server environment, and swagger-tools is not involved, we don't document it.  Can you tell me what is going wrong...maybe with a little more of your server initialization code so I can see the order of `app.use` and such?

**Comment by rw12345** [comment]
Thanks for the quick reply. 
Here's what the server initialization looks like. As you can see, steps 1 to 7 are fairly standard and works nicely. It's just 8 that I'm not sure about.

`server.js`
```
// Dependency
var SwaggerParser = require('swagger-parser');
var SwaggerExpress = require('swagger-express-mw');
var SwaggerTools = require('swagger-tools');
var cors = require('cors');

...
// swagger router config
// See: https://github.com/apigee-127/swagger-tools/blob/master/docs/Middleware.md
var swaggerRouterOptions = {
  controllers: './api/controllers',
  useStubs: process.env.NODE_ENV === 'development'
};
...

// Server._init is called to initialize the middleware etc.
Server.prototype._init = function(callback) {
  var _this = this;

  // 1) Load the Swagger spec
  SwaggerParser.parse('./api/swagger/swagger.yaml', function(error, api) {
    if (error) return callback(error);   
   
   // 2) Create the express app
   _this.app = require('express')();

    // 3) Initialize Swagger Tools
    SwaggerTools.initializeMiddleware(api, function(middleware) {      
      _this.app.use(middleware.swaggerMetadata());

      // 4) Hook up validator
      _this.app.use(middleware.swaggerValidator({ validateResponse: true }));

      // 5) Route validated requests to appropriate controller
      _this.app.use(middleware.swaggerRouter(swaggerRouterOptions));

      // 6) Hook up Swagger express middleware
      SwaggerExpress.create(swaggerExpressConfig, function(error, swaggerExpress) {
        if (error) return callback(error);
        
        swaggerExpress.register(_this.app);

        // 7) Adding global middleare works fine. For example...
       _this.app.use(cors());
       
        // 8) But how would you hook up route specific middleware as swaggerRouter has already loaded and initialized the routes?
        // In theory, it should be something like:

        
        // 9) All complete
        return callback(null);
      });
    });
  });
};

```

**Comment by whitlockjc** [comment]
I think using swagger-tools and swagger-express-mw at the same time is redundant.  Basically, all of the middlewares you wire up from swagger-tools will be run twice because they are also wired up in swagger-express-mw as well.

As for your problem, running step 8 after step 5 means that swagger-router will **always** run prior to your custom route specific handlers so those would need to be wired up prior to step 5.  I'm sure you've tried this so what happened?  We can work around swagger-router if the need arises but what happened that you didn't expect when you tried?  *(Realize that some of this not working might be using swagger-tools and swagger-express-mw together.)*

**Comment by rw12345** [comment]
@whitlockjc Apologies for the late response and the good news is, you are absolutely correct. 
I missed initialising the route handlers before initialising the swagger-router, which obviously led to the weiredness I was seeing. Doh!
Many thanks for the clarification.

**Comment by sandip8600** [comment]
@whitlockjc  How did you add route specific middleware. can you post snippet ?

**Comment by whitlockjc** [comment]
It's no different than any other middleware so it should be something like this:

```js
// ...
app.get("/some/path", middleware.swaggerRouter);
// ...
```

I've not tested this but just know that swagger-tools' middleware is no different than any other Connect-based middleware.  That means however Connect or Express or whomever would do it, use that same approach.

**Comment by Maples7** [comment]
@whitlockjc 
I've read the source code briefly, and maybe you can make handlers support both single function and array of functions?

for example, while `operationId` is `createPuppy` and `createPuppy` is acutally an array of functions, execute them according to the order in the array.

Just like this for `controllers`,
```js
module.exports.createPuppy = [
  (req, res, next) => {
    res.myOwnVar = 'this is a middleware.';
    next();
  },
  (req, res, next) => {
    res.end('success');
  }
]
```
Would this be better?

**Comment by whitlockjc** [comment]
Well, I don't want `swagger-tools` becoming the middleware provider.  Why doesn't the standard Connect-based way work?

**Comment by Maples7** [comment]
With swagger-tools, routes and controllers are parsed from the doc definition directly, which changes connect/express standard ways of processing pipeline, so maybe it's a better idea to also make this package support middlewares. By doing so, the pipeline would be very clear.

Just my thought, and thanks for reply.

**Comment by whitlockjc** [comment]
What I'm asking is if you look at the responses above, it appears you can already do this.

**Comment by laterne** [comment]
I'm running into the same or similar issue. I'll try to illustrate it with the following example:
The great advantage of something like swaggerRouter in combination with connect/express is that I don't have to replicate the path declaration from the swagger file in my code:
```
paths:
/users/{user_id}:
    $ref: "paths/users.yaml"
/ponies/{pony_id}:
    $ref: "paths/ponies.yaml"
```
declaration of route here is delegated to swaggerRouter and also changed easily because I only need to change urls once in the swaggerfiles
```javascript
app.get('/users', some, middleware, users.sendResponse)
app.get('/users/:userid', more, userMiddleware, user.sendResponse)
app.get('/ponies/:ponyid', different, ponyMidleware, pony.sendResponse)
```
It would still be nice to have the different stacks of middleware `some, middleware, users.sendResponse`,  `different, ponyMidleware, pony.sendResponse` to be called when the swaggerRouter matches the specific route.

@whitlockjc In your example, `app.get("/some/path", middleware.swaggerRouter);`, it is only possible to mount middleware before the router and for all paths.

It would be great if swaggerRouter could not only call a single middleware `function(req, res, next)` that handles the route but also a stack of middleware `var app = connect(); app.use(middleware)`. Mayby @Maples7 meant  this, because in express it is possible to mount multiple middleware functions at the same time in the form of `app.use([some, reusable, middleware])`.
With swaggerRouter calling a stack of middleware it would not be a middleware provider and still be server agnostic.

May be this could also be considered for [sway-connect](https://github.com/whitlockjc/sway-connect/issues/3)?

**Comment by whitlockjc** [comment]
I'm not understanding the issue here.  All `swagger-tools` middleware will process all paths and respond accordingly, just like any other middleware out there.  If you only want it to register `swagger-tools` middlewares for certain paths instead of a _global_ registration, why can't you use `app.[verb](path, swaggerTools.someMiddleware)` instead of `app.use(swaggerTools.someMiddleware)`?

**Comment by SalathielGenese** [comment]
Hi guys,

I'm still unsure why some see the problem as solved, but... This' how I'll handle it :

```js
const { jwtMiddleware } = require('./path/to/middleware.jwt');


/**
 * @swagger
 * My SwaggerJSDoc config I coupled with swagger-tools
 *
 **/
const criticalAction = (request, response) => jwtMiddleware(request, response, () => {
    // my ordinary express code
});




module.exports = {
    criticalAction,
}
```;

**Comment by Maples7** [comment]
It's all about [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and how to make this tool  handier. And supporting this feature is painless as what I recall in the source code.
